### PR TITLE
Make dump print `const` before const fields

### PR DESCRIPTION
**Pull request by jakobnissen**
This provides more information to the user when dumping types, and also makes the output of dump slightly more similar to the type definition syntax.

EDIT: This has been changed to print:
* The kind of type before the type name `abstract type`, `mutable struct`, etc.
* `const` only for `const` fields of `mutable struct`

New behaviour
```
julia> dump(Float32)
primitive type Float32 <: AbstractFloat

julia> dump(Signed)
abstract type Signed <: Integer

julia> dump(Pair{Int, String})
struct Pair{Int64, String} <: Any
  first::Int64
  second::String

julia> dump(BitSet)
mutable struct BitSet <: AbstractSet{Int64}
  const bits::Vector{UInt64}
  offset::Int64

julia> dump(Set)
UnionAll
  var: TypeVar
    name: Symbol T
    lb: Union{}
    ub: abstract type Any
  body: struct Set{T} <: AbstractSet{T}
    dict::Dict{T, Nothing}
```

### CODE REVIEW

**Review by oscardssmith — APPROVED** [review]

**Comment by IanButterworth** [comment]
```
  | show                                             (9) \|         failed at 2024-02-27T05:53:21.654
  | Test Failed at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-HL2F7YQ3XH.0/build/default-honeycrisp-HL2F7YQ3XH-0/julialang/julia-master/julia-64875222df/share/julia/test/show.jl:1266
  | Expression: repr == "Pair{String, Int64} <: Any\n  first::String\n  second::Int64\n"
  | Evaluated: "Pair{String, Int64} <: Any\n  const first::String\n  const second::Int64\n" == "Pair{String, Int64} <: Any\n  first::String\n  second::Int64\n"
```

File: test/show.jl [hunk 1]
```diff
@@ -1263,7 +1263,7 @@ let repr = sprint(dump, :(x = 1))
     @test repr == "Expr\n  head: Symbol =\n  args: Array{Any}((2,))\n    1: Symbol x\n    2: $Int 1\n"
 end
 let repr = sprint(dump, Pair{String,Int64})
-    @test repr == "Pair{String, Int64} <: Any\n  first::String\n  second::Int64\n"
```
Comment (KristofferC) [inline test/show.jl:1266]: So all non mutable structs will print `const` for all their fields? Isn't it better to leave the non-mutable structs as is and only print out `const` for mutable structs with const fields (together with adding `mutable` in the dump for those types).

Comment (jakobnissen) [reply test/show.jl:1266]: You're right, that might be strange. I've now made the following changes:
* Print the type of type (`struct`, `primitive type` etc.) at the beginning before the name
* Don't print `const` unless in a mutable struct.

Here are some examples:
```
julia> dump(Float32)
primitive type Float32 <: AbstractFloat

julia> dump(Signed)
abstract type Signed <: Integer

julia> dump(Pair{Int, String})
struct Pair{Int64, String} <: Any
  first::Int64
  second::String

julia> dump(BitSet)
mutable struct BitSet <: AbstractSet{Int64}
  const bits::Vector{UInt64}
  offset::Int64

julia> dump(Set)
UnionAll
  var: TypeVar
    name: Symbol T
    lb: Union{}
    ub: abstract type Any
  body: struct Set{T} <: AbstractSet{T}
    dict::Dict{T, Nothing}
```

Comment (KristofferC) [reply test/show.jl:1266]: 👌

**Pull request merged**